pkg/build/controller/build/build_controller.go

```diff
@@ -1090,10 +1090,11 @@ func (bc *BuildController) handleActiveBuild(build *buildv1.Build, pod *corev1.P
 	podPhase := pod.Status.Phase
 	var update *buildUpdate
 	// Pods don't report running until initcontainers are done, but from a build's perspective
-	// the pod is running as soon as the first init container has run.
+	// the pod is running as soon as the first init container is running or terminated.
+	// note, the git-clone could be terminated, but a subsequent init container could be running
 	if build.Status.Phase == buildv1.BuildPhasePending || build.Status.Phase == buildv1.BuildPhaseNew {
 		for _, initContainer := range pod.Status.InitContainerStatuses {
-			if initContainer.Name == strategy.GitCloneContainer && initContainer.State.Running != nil {
+			if initContainer.Name == strategy.GitCloneContainer && (initContainer.State.Running != nil || initContainer.State.Terminated != nil) {
```

Comment:
the bug is that we attempt to transition the build phase from "running" to "pending" but this if block will only be entered if the build phase is "pending" or "new", so how can this be the cause of the bug?

Reply:
yep change is incomplete / got turned around a bit with different versions of my work here  ... this specific change helps get it to running if something happened to the event where `initContainer.State.Running != nil` was missed, but we then got the `initContainer.State.Terminated != nil` event.

Will push subsequent change up in a bit.

Reply:
At this point, given the difficulty in getting diagnostics from the customer, I plan on putting some better defensive programming in at this point.

Reply:
Per [k8s API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#containerstate-v1-core):

> ContainerState holds a possible state of container. Only one of its members may be specified. If none of them is specified, the default one is ContainerStateWaiting.

So the build pod can start with the status "Running" from the git clone. It then moves to "Pending" because we aren't checking the terminated status.

Reply:
We also have init containers to extract image content and manage the Dockerfile. If either of these are slow (possible if a lot of image content is being extracted) then I can see why this would show up as a "smell" for an otherwise slow build.

Reply:
> So the build pod can start with the status "Running" from the git clone. It then moves to "Pending" because we aren't checking the terminated status.

A pod can't move from running to pending as far as i know:

a pod will go through:
0) pending
1) init (init containers running)
2) running
3) terminated

and for the build itself, it stops caring about the init container state as soon as it gets to "running" which it gets to as soon as it sees the gitclone container run.

But yes, it's possible (prior to Gabe's fix here) for a build to appear to sit in pending longer than it should because the gitclone init container runs quickly, and then a subsequent init container runs slowly, and we miss the update that showed us the gitclone container was running.  Won't cause a bad transition, but will cause what appears to be a "stuck" build until the init container finishes.

Reply:
you can check with @sjenning as to whether a pod can transition back to "pending" after completing init containers and before starting the main container, but i don't think that happens.